File: 20242Q/poo/laboratorio/tp05/_soluciones/ej3/figures/Triangle.java
package ar.edu.itba.poo.soltp5.ej3.figures;

public class Triangle<P extends Point> extends Figure {

    protected final P firstPoint, secondPoint, thirdPoint;

    public Triangle(P firstPoint, P secondPoint, P thirdPoint) {
        this.firstPoint = firstPoint;
        this.secondPoint = secondPoint;
        this.thirdPoint = thirdPoint;
    }

    @Override
    public double area() {
        /**
         * Como desconocemos la base y la altura del triángulo
         * usamos la fórmula de Heron que utiliza el semiperímetro
         */
        double semiPerimiter = perimeter() / 2;
        return Math.sqrt(semiPerimiter * (semiPerimiter - firstSide())
                * (semiPerimiter - secondSide()) * (semiPerimiter - thirdSide()));
    }

    @Override
    public double perimeter() {
        return firstSide() + secondSide() + thirdSide();
    }

    private double firstSide() {
        return firstPoint.distanceTo(secondPoint);
    }

    private double secondSide() {
        return secondPoint.distanceTo(thirdPoint);
    }

    private double thirdSide() {
        return thirdPoint.distanceTo(firstPoint);
    }

    @Override
    public String toString() {
        return "Triángulo [ %s , %s , %s ]".formatted(firstPoint, secondPoint, thirdPoint);
    }

}

File: 20242Q/poo/laboratorio/tp08/_soluciones/ej4/triangle.rb
require_relative 'figure'

class Triangle < Figure
  def initialize(first_point, second_point, third_point)
    @first_point = first_point
    @second_point = second_point
    @third_point = third_point
  end

  def area
    semiperimiter = (first_side + second_side + third_side) / 2
    Math.sqrt(semiperimiter * (semiperimiter - first_side) * (semiperimiter - second_side) * (semiperimiter - third_side))
  end

  def perimeter
    first_side + second_side + third_side
  end

  def to_s
    "Triángulo [ #{@first_point} , #{@second_point} , #{@third_point} ]"
  end

  private

  def first_side
    @first_point.distance(@second_point)
  end

  def second_side
    @second_point.distance(@third_point)
  end

  def third_side
    @third_point.distance(@first_point)
  end
end
